File: yahtzee_scala_analysis/performance_profiler.scala
package analysis

import yahtzee.*
import yahtzee.Dice.*
import scala.util.Using
import java.io.PrintWriter
import scala.collection.mutable.ArrayBuffer

case class PerformanceMetrics(
  executionTimeMs: Double,
  memoryUsedKb: Double
)

object PerformanceProfiler:
  
  def profileMethod[T](name: String, iterations: Int = 1000)(f: => T): PerformanceMetrics =
    // Warmup
    (0 until 100).foreach(_ => f)
    
    // Force GC before measurement
    System.gc()
    Thread.sleep(50)
    
    val runtime = Runtime.getRuntime
    val memBefore = runtime.totalMemory() - runtime.freeMemory()
    
    val startTime = System.nanoTime()
    (0 until iterations).foreach(_ => f)
    val endTime = System.nanoTime()
    
    val memAfter = runtime.totalMemory() - runtime.freeMemory()
    
    val avgTimeMs = (endTime - startTime) / iterations / 1e6
    val memUsedKb = Math.max(0, memAfter - memBefore) / 1024.0
    
    PerformanceMetrics(avgTimeMs, memUsedKb)
  
  def profileDice(): Map[String, PerformanceMetrics] =
    val results = scala.collection.mutable.Map[String, PerformanceMetrics]()
    var seed = 12345L
    
    // Dice creation
    results("dice_create") = profileMethod("dice_create") {
      val (d, s) = Dice.create(seed)
      seed = s
      d
    }
    
    // Dice apply
    results("dice_apply") = profileMethod("dice_apply") {
      Dice(3)
    }
    
    // Dice value access
    val dice = Dice(3)
    results("dice_value") = profileMethod("dice_value") {
      dice.value
    }
    
    // Dice roll
    results("dice_roll") = profileMethod("dice_roll") {
      val (d, s) = dice.roll(seed)
      seed = s
      d
    }
    
    results.toMap
  
  def profileScorecard(): Map[String, PerformanceMetrics] =
    val results = scala.collection.mutable.Map[String, PerformanceMetrics]()
    
    // Scorecard creation
    results("scorecard_apply") = profileMethod("scorecard_apply") {
      Scorecard(2)
    }
    
    val scorecard = Scorecard(2)
    
    // Get score
    results("scorecard_getScore") = profileMethod("scorecard_getScore") {
      scorecard.getScore(0, 0)
    }
    
    // Set score
    results("scorecard_setScore") = profileMethod("scorecard_setScore") {
      scorecard.setScore(0, 0, 10)
    }
    
    // Get player card
    results("scorecard_getPlayerCard") = profileMethod("scorecard_getPlayerCard") {
      scorecard.getPlayerCard(0)
    }
    
    results.toMap
  
  def profileGame(): Map[String, PerformanceMetrics] =
    val results = scala.collection.mutable.Map[String, PerformanceMetrics]()
    var seed = 12345L
    
    // Create test dice
    val (dice, newSeed) = (0 until 5).foldLeft((List.empty[Dice], seed)) { case ((acc, s), _) =>
      val (d, ns) = Dice.create(s)
      (acc :+ d, ns)
    }
    seed = newSeed
    
    // Roll dice
    results("game_rollDice") = profileMethod("game_rollDice") {
      val (d, s) = Game.rollDice(dice, seed)
      seed = s
      d
    }
    
    // Roll specific dice
    results("game_rollSpecificDice") = profileMethod("game_rollSpecificDice") {
      val (d, s) = Game.rollSpecificDice(dice, Set(0, 2, 4), seed)
      seed = s
      d
    }
    
    // Get dice values
    results("game_getDiceValues") = profileMethod("game_getDiceValues") {
      Game.getDiceValues(dice)
    }
    
    // Get sorted dice
    results("game_getSortedDice") = profileMethod("game_getSortedDice") {
      Game.getSortedDice(dice)
    }
    
    // Get frequency
    results("game_getFrequency") = profileMethod("game_getFrequency") {
      Game.getFrequency(dice)
    }
    
    // Calculate scores for all categories
    (0 until 13).foreach { slot =>
      val categoryName = slot match
        case 0 => "ones"
        case 1 => "twos"
        case 2 => "threes"
        case 3 => "fours"
        case 4 => "fives"
        case 5 => "sixes"
        case 6 => "three_of_kind"
        case 7 => "four_of_kind"
        case 8 => "full_house"
        case 9 => "small_straight"
        case 10 => "large_straight"
        case 11 => "yahtzee"
        case 12 => "chance"
      
      results(s"game_calculateScore_$categoryName") = profileMethod(s"game_calculateScore_$categoryName") {
        Game.calculateScore(dice, slot)
      }
    }
    
    // Straight detection
    val sortedDice = List(1, 2, 3, 4, 5)
    results("game_isSmallStraight") = profileMethod("game_isSmallStraight") {
      Game.isSmallStraight(sortedDice)
    }
    
    results("game_isLargeStraight") = profileMethod("game_isLargeStraight") {
      Game.isLargeStraight(sortedDice)
    }
    
    results.toMap
  
  def profileGameState(): Map[String, PerformanceMetrics] =
    val results = scala.collection.mutable.Map[String, PerformanceMetrics]()
    var seed = 12345L
    
    // GameState initial
    results("gamestate_initial") = profileMethod("gamestate_initial") {
      GameState.initial(2, seed)
    }
    
    val state = GameState.initial(2, seed)
    
    // Roll dice
    results("gamestate_rollDice") = profileMethod("gamestate_rollDice") {
      state.rollDice(Set(0, 2, 4))
    }
    
    // Record score
    results("gamestate_recordScore") = profileMethod("gamestate_recordScore") {
      state.recordScore(0)
    }
    
    // Is game over
    results("gamestate_isGameOver") = profileMethod("gamestate_isGameOver") {
      state.isGameOver
    }
    
    results.toMap
  
  def toJson(allResults: Map[String, Map[String, PerformanceMetrics]]): String =
    val sb = new StringBuilder
    sb.append("{\n")
    
    val modules = allResults.toList
    modules.zipWithIndex.foreach { case ((moduleName, methods), moduleIdx) =>
      sb.append(s"""  "$moduleName": {\n""")
      val methodList = methods.toList
      methodList.zipWithIndex.foreach { case ((methodName, metrics), methodIdx) =>
        sb.append(s"""    "$methodName": {\n""")
        sb.append(s"""      "execution_time_ms": ${metrics.executionTimeMs},\n""")
        sb.append(s"""      "memory_used_kb": ${metrics.memoryUsedKb}\n""")
        sb.append(s"""    }${if methodIdx < methodList.size - 1 then "," else ""}\n""")
      }
      sb.append(s"""  }${if moduleIdx < modules.size - 1 then "," else ""}\n""")
    }
    
    sb.append("}\n")
    sb.toString
  
  def main(args: Array[String]): Unit =
    println("Running Scala performance analysis...")
    
    val allResults = Map(
      "dice" -> profileDice(),
      "scorecard" -> profileScorecard(),
      "game" -> profileGame(),
      "gamestate" -> profileGameState()
    )
    
    // Save to JSON
    val jsonOutput = toJson(allResults)
    val outputPath = "../yahtzee_scala_analysis/data/performance_metrics.json"
    Using(new PrintWriter(outputPath)) { writer =>
      writer.write(jsonOutput)
    }
    
    // Print results
    println("\nPerformance Analysis Results:")
    println("=" * 80)
    
    allResults.foreach { case (module, methods) =>
      println(s"\n${module.toUpperCase}")
      println("-" * 80)
      methods.toList.sortBy(_._1).foreach { case (method, metrics) =>
        println(f"  $method%-35s | Time: ${metrics.executionTimeMs}%8.4f ms | Memory: ${metrics.memoryUsedKb}%8.2f KB")
      }
    }
    
    println(s"\nResults saved to: $outputPath")

File: yahtzee_analysis/performance_profiler.py
import time
import tracemalloc
import json
from pathlib import Path
import sys

yahtzee_game_path = Path(__file__).parent.parent / 'yahtzee_game'
sys.path.insert(0, str(yahtzee_game_path))

from dice import Dice
from scorecard import Scorecard
from game import Game

def profile_method(func, *args, **kwargs):
    tracemalloc.start()
    start_time = time.perf_counter()
    start_memory = tracemalloc.get_traced_memory()[0]
    
    result = func(*args, **kwargs)
    
    end_time = time.perf_counter()
    current_memory, peak_memory = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    return {
        'execution_time_ms': (end_time - start_time) * 1000,
        'memory_used_kb': (current_memory - start_memory) / 1024,
        'peak_memory_kb': peak_memory / 1024
    }

def profile_dice():
    results = {}
    
    dice = Dice()
    results['dice_init'] = profile_method(Dice)
    results['dice_roll'] = profile_method(dice.roll)
    results['dice_face_value_get'] = profile_method(lambda: dice.face_value)
    results['dice_face_value_set'] = profile_method(setattr, dice, '_face_value', 3)
    
    return results

def profile_scorecard():
    results = {}
    
    results['scorecard_init'] = profile_method(Scorecard, 2)
    scorecard = Scorecard(2)
    results['scorecard_get_score'] = profile_method(scorecard.get_score, 0, 0)
    results['scorecard_set_score'] = profile_method(scorecard.set_score, 0, 0, 10)
    results['scorecard_get_player_card'] = profile_method(scorecard.get_player_card, 0)
    
    return results

def profile_game():
    results = {}
    
    # Initialization
    results['game_init'] = profile_method(Game, 2)
    game = Game(2)
    
    # Dice rolling and value management
    results['game_roll_dice'] = profile_method(game.roll_dice)
    results['game_get_dice_values'] = profile_method(game.get_dice_values)
    results['game_set_dice_values'] = profile_method(game.set_dice_values)
    
    # Sorted dice operations
    results['game_get_sorted_dice'] = profile_method(game.get_sorted_dice)
    results['game_set_sorted_dice'] = profile_method(game.set_sorted_dice)
    
    # Frequency operations
    results['game_get_frequency'] = profile_method(game.get_frequency)
    results['game_set_frequency'] = profile_method(game.set_frequency)
    
    # Scoring calculations (all categories)
    results['game_calculate_score_ones'] = profile_method(game.calculate_score, 0)
    results['game_calculate_score_twos'] = profile_method(game.calculate_score, 1)
    results['game_calculate_score_threes'] = profile_method(game.calculate_score, 2)
    results['game_calculate_score_fours'] = profile_method(game.calculate_score, 3)
    results['game_calculate_score_fives'] = profile_method(game.calculate_score, 4)
    results['game_calculate_score_sixes'] = profile_method(game.calculate_score, 5)
    results['game_calculate_score_three_of_kind'] = profile_method(game.calculate_score, 6)
    results['game_calculate_score_four_of_kind'] = profile_method(game.calculate_score, 7)
    results['game_calculate_score_full_house'] = profile_method(game.calculate_score, 8)
    results['game_calculate_score_small_straight'] = profile_method(game.calculate_score, 9)
    results['game_calculate_score_large_straight'] = profile_method(game.calculate_score, 10)
    results['game_calculate_score_yahtzee'] = profile_method(game.calculate_score, 11)
    results['game_calculate_score_chance'] = profile_method(game.calculate_score, 12)
    
    # Straight detection helpers
    results['game_is_small_straight'] = profile_method(game.is_small_straight, [1, 2, 3, 4, 5])
    results['game_is_large_straight'] = profile_method(game.is_large_straight, [1, 2, 3, 4, 5])
    
    # Display methods (using StringIO to capture output without printing)
    from io import StringIO
    import sys
    old_stdout = sys.stdout
    sys.stdout = StringIO()
    try:
        results['game_display_dice'] = profile_method(game.display_dice, set())
        results['game_display_scorecard'] = profile_method(game.display_scorecard, 0)
    finally:
        sys.stdout = old_stdout
    
    return results

def run_performance_analysis():
    print("Running performance analysis...")
    
    all_results = {
        'dice': profile_dice(),
        'scorecard': profile_scorecard(),
        'game': profile_game()
    }
    
    output_dir = Path(__file__).parent / 'data'
    output_dir.mkdir(exist_ok=True)
    
    with open(output_dir / 'performance_metrics.json', 'w') as f:
        json.dump(all_results, f, indent=2)
    
    print("\nPerformance Analysis Results:")
    print("=" * 70)
    for module, methods in all_results.items():
        print(f"\n{module.upper()}")
        print("-" * 70)
        for method, metrics in methods.items():
            print(f"  {method:30s} | Time: {metrics['execution_time_ms']:8.4f}ms | Memory: {metrics['memory_used_kb']:8.2f}KB")
    
    print(f"\nResults saved to: {output_dir / 'performance_metrics.json'}")
    return all_results

if __name__ == '__main__':
    run_performance_analysis()
